File: benchmarks/js/call/method_dispatch.js
// Method dispatch benchmark
// Tests prototype-based OOP: class creation, object instantiation, method calls
// Equivalent to Lua metatable-based dispatch

var print = typeof console !== 'undefined' ? console.log.bind(console) : globalThis.print;
var DEFAULT = 5000000;
var N = parseInt(typeof scriptArgs !== 'undefined' ? scriptArgs[1] : (typeof process !== 'undefined' ? process.argv[2] : '0')) || DEFAULT;

// Base class
function Animal(name, legs, sound) {
    this.name = name;
    this.legs = legs;
    this.sound = sound;
    this.energy = 100;
}

Animal.prototype.speak = function() {
    return this.sound;
};

Animal.prototype.move = function(dist) {
    this.energy = this.energy - dist;
    return this.energy;
};

Animal.prototype.rest = function(amount) {
    this.energy = this.energy + amount;
    return this.energy;
};

// Derived class: Dog
function Dog(name) {
    Animal.call(this, name, 4, "woof");
}
Dog.prototype = Object.create(Animal.prototype);
Dog.prototype.constructor = Dog;

Dog.prototype.fetch = function(dist) {
    this.move(dist);
    this.move(dist);
    return this.energy;
};

// Derived class: Bird
function Bird(name) {
    Animal.call(this, name, 2, "tweet");
}
Bird.prototype = Object.create(Animal.prototype);
Bird.prototype.constructor = Bird;

Bird.prototype.fly = function(dist) {
    this.move(dist * 2);
    return this.energy;
};

// Create objects and call methods polymorphically
var animals = [
    new Dog("Rex"),
    new Bird("Tweety"),
    new Dog("Buddy"),
    new Bird("Eagle"),
    new Animal("Cat", 4, "meow"),
];

var total = 0;
var nanimals = animals.length;

for (var i = 1; i <= N; i++) {
    var a = animals[i % nanimals];
    a.rest(10);
    a.move(3);
    total += a.energy;
    a.speak();
}

print(total);

File: benchmarks/lua/call/method_dispatch.lua
-- Method dispatch benchmark
-- Tests metatable-based OOP: class creation, object instantiation, method calls
-- Representative of production Lua patterns (OpenResty, game scripting)

local N = tonumber(arg and arg[1]) or 5000000
-- collectgarbage("collect")

-- Base class via metatables
local Animal = {}
Animal.__index = Animal

function Animal.new(name, legs, sound)
    local self = setmetatable({}, Animal)
    self.name = name
    self.legs = legs
    self.sound = sound
    self.energy = 100
    return self
end

function Animal:speak()
    return self.sound
end

function Animal:move(dist)
    self.energy = self.energy - dist
    return self.energy
end

function Animal:rest(amount)
    self.energy = self.energy + amount
    return self.energy
end

-- Derived class
local Dog = setmetatable({}, { __index = Animal })
Dog.__index = Dog

function Dog.new(name)
    local self = Animal.new(name, 4, "woof")
    return setmetatable(self, Dog)
end

function Dog:fetch(dist)
    self:move(dist)
    self:move(dist)
    return self.energy
end

-- Another derived class
local Bird = setmetatable({}, { __index = Animal })
Bird.__index = Bird

function Bird.new(name)
    local self = Animal.new(name, 2, "tweet")
    return setmetatable(self, Bird)
end

function Bird:fly(dist)
    self:move(dist * 2)
    return self.energy
end

-- Create objects and call methods polymorphically
local animals = {
    Dog.new("Rex"),
    Bird.new("Tweety"),
    Dog.new("Buddy"),
    Bird.new("Eagle"),
    Animal.new("Cat", 4, "meow"),
}

local total = 0
local nanimals = #animals

for i = 1, N do
    local a = animals[(i % nanimals) + 1]
    a:rest(10)
    a:move(3)
    total = total + a.energy
    a:speak()
end

print(total)
